dir/fs.go

```diff
@@ -0,0 +1,159 @@
+package dir
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+// rootedFS is io.FS implementation used in New.
+// root is the root of the file system tree passed to os.DirFS.
+type rootedFS struct {
+	fs.FS
+	Root string
+}
+
+// NewRootedFS create a rootedFS
+func NewRootedFS(fsys fs.FS, root string) rootedFS {
+	return rootedFS{fsys, root}
+}
+
+// UnionDirFS is a simple union directory file system interface
+type UnionDirFS interface {
+	fs.ReadDirFS
+	GetPath(elem ...string) (string, error)
+}
+
+// NewUnionDirFS create an unionDirFS on the current OS file system
+// by the named dirs
+func NewUnionDirFS(dirs ...string) unionDirFS {
+	rootedDirs := []rootedFS{}
+	for _, dir := range dirs {
+		rootedDirs = append(rootedDirs, rootedFS{
+			FS:   os.DirFS(dir),
+			Root: dir,
+		})
+	}
+	return unionDirFS{Dirs: rootedDirs}
+}
+
+// NewUnionDirFsByRootedFS create an unionDirFS by rootedFS
+func NewUnionDirFsByRootedFS(rootedFsys ...rootedFS) unionDirFS {
+	return unionDirFS{Dirs: rootedFsys}
+}
+
+// unionDirFS is a simple union directory file system
+//
+// it unions multiple directory to be one directory with priority based on the
+// order of Dirs
+type unionDirFS struct {
+	Dirs []rootedFS
+}
+
+// Open implements fs.FS interface
+//
+// traverse all union directories and return the first existing file
+func (u unionDirFS) Open(name string) (fs.File, error) {
+	for _, dir := range u.Dirs {
+		f, err := dir.Open(name)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+		return f, nil
+	}
+	return nil, &fs.PathError{Op: "open", Err: fs.ErrNotExist, Path: name}
+}
+
+// GetPath returns the path of the named file or directory under the FS
+//
+// if path exists, it returns the first existing path in union directories (dirs)
+// if path doesn't exist, it returns fs.ErrNotExist error
+func (u unionDirFS) GetPath(elem ...string) (string, error) {
+	var targetPath string
+	pathSuffix := path.Join(elem...)
+	for _, dir := range u.Dirs {
+		targetPath = filepath.Join(dir.Root, pathSuffix)
+		statDir, ok := dir.FS.(fs.StatFS)
+		if !ok {
+			return "", &fs.PathError{
+				Op:  "getpath",
+				Err: errors.New("not implemented"),
+			}
+		}
+		_, err := statDir.Stat(pathSuffix)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				continue
+			} else {
+				return "", err
+			}
+		}
+		// got the first existing file path and break
+		// return the path with current OS separator
+		return targetPath, nil
```

Comment:
When I say `fs.Stat`, I mean [fs.Stat()](https://pkg.go.dev/io/fs#Stat). The built-in library will take care of everything.

Reply:
Good idea! Thanks.